File: Ch02/ex2_05.erl
-module(ex2_05).
-author("Kevin C. Baird").
-purpose("SICP in Erlang: Exercise 2.5 - cons, car, cdr via 2 ** a * 3 ** b.").
-export([cons/2, car/1, cdr/1]).

cons(A, B) -> trunc(math:pow(2, A) * math:pow(3, B)).
car(Z)     -> car(Z, 0).
cdr(Z)     -> cdr(Z, 0).

% helpers
car(A, B) ->
  case A rem 2 of
    0 -> car((A div 2), (B+1));
    _ -> B
  end.

cdr(A, B) ->
  case A rem 3 of
    0 -> cdr((A div 3), (B+1));
    _ -> B
  end.

File: Ch02/ex2_05.exs
# SICP in Elixir: Exercise 2.5 - cons, car, cdr via 2 ** a * 3 ** b

defmodule Ex2_05 do
  def cons(a, b) do
    trunc(:math.pow(2, a) * :math.pow(3, b))
  end

  def car(z) do
    extract(z, 0, 2)
  end

  def cdr(z) do
    extract(z, 0, 3)
  end

  # Private functions
  defp extract(encoded, candidate, base) do
    case rem(encoded, base) do
      0 -> extract(div(encoded, base), candidate + 1, base)
      _ -> candidate
    end
  end
end

list = Ex2_05.cons(1, 2)
IO.puts(Ex2_05.car(list))
IO.puts(Ex2_05.cdr(list))
